pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,484 @@
+
+		t.nodesLock.Lock()
+		defer t.nodesLock.Unlock()
+
+		for nodeID, dest := range t.nodes {
+			// Close an old connection to a node if we don't need to connect to it further.
+			if _, foundInNewNodes := newNodes[nodeID]; !foundInNewNodes {
+				t.logger.Log(logging.LevelDebug, "Closing old connection", "src", t.ownID, "dst", nodeID)
+
+				if err := t.host.Network().ClosePeer(dest.AddrInfo.ID); err != nil {
+					t.logger.Log(logging.LevelError, "Could not close old connection to node", "src", t.ownID, "dst", nodeID, "err", err)
+					continue
+				}
+
+				delete(t.nodes, nodeID)
+			}
+		}
+	}()
+}
+
+func (t *Transport) Connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
+	if len(nodes) == 0 {
+		t.logger.Log(logging.LevelWarn, "no nodes to connect to")
+		return
+	}
+
+	t.nodesLock.Lock()
+	for nodeID, addr := range nodes {
+		if nodeID == t.ownID {
+			continue
+		}
+		_, found := t.nodes[nodeID]
+		if !found {
+			info, err := peer.AddrInfoFromP2pAddr(addr)
+			if err != nil {
+				t.logger.Log(logging.LevelWarn, "connect: failed to parse addr", "src", t.ownID, "dest", nodeID, "addr", addr, "err", err)
+				continue
+			}
+			t.nodes[nodeID] = &nodeInfo{
+				ID:        nodeID,
+				Addr:      addr,
+				AddrInfo:  info,
+				IsOpening: false,
+				Stream:    nil,
+			}
+
+		}
+	}
+	t.nodesLock.Unlock()
+
+	t.connect(ctx, nodes)
+}
+
+func (t *Transport) connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
+	t.connWg.Add(1)
+	defer t.connWg.Done()
+
+	go func() {
+		t.logger.Log(logging.LevelDebug, "started connecting nodes", "src", t.ownID)
+		defer t.logger.Log(logging.LevelDebug, "finished connecting nodes", "src", t.ownID)
+
+		wg := &sync.WaitGroup{}
+		wg.Add(len(nodes))
+
+		for nodeID := range nodes {
+			if nodeID == t.ownID {
+				// Do not establish a real connection with own node.
+				wg.Done()
+				continue
+			}
+
+			if t.streamExists(nodeID) {
+				t.logger.Log(logging.LevelInfo, "stream to node already exists", "src", t.ownID, "dst", nodeID)
+				wg.Done()
+				continue
+			}
+
+			go t.connectToNode(ctx, nodeID, wg)
+		}
+		wg.Wait()
+	}()
+}
+
+func (t *Transport) reconnect(ctx context.Context, node types.NodeID) {
+	info, found := t.getNodeAddrInfo(node)
+	if !found {
+		t.logger.Log(logging.LevelError, "failed to get node address", "src", t.ownID, "dst", node)
+	} else {
+		if err := t.host.Network().ClosePeer(info.ID); err != nil {
+			t.logger.Log(logging.LevelError, "could not close the connection to node", "src", t.ownID, "dst", node, "err", err)
+		}
+		t.connWg.Add(1)
+		go t.connectToNode(ctx, node, t.connWg)
+	}
+}
+
+func (t *Transport) connectToNode(ctx context.Context, node types.NodeID, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if t.isConnectionToNodeInProgress(node) {
+		t.logger.Log(logging.LevelDebug, "connecting to node is in progress", "src", t.ownID, "dst", node)
+		return
+	}
+
+	t.setConnectionInProgress(node)
+	defer t.clearConnectionInProgress(node)
```

Comment:
Are those calls supposed to make some synchronization scheme up or they are just there for reducing the likelihood of duplicate attempt to establish the same connection? I think there is a possible race condition since `t.nodesLock` gets unlocked in-between.

Reply:
Fixed